Fix the -R option on Debian/Ubuntu

rallytime opened the pull request:
### What does this PR do?
When the `-R` option was originally implemented for Debian/Ubuntu, I must not have been testing the right thing because I am not sure how that ever could have worked. This PR reworks some of those `-R` checks and sets up the repositories correctly.

### What issues does this PR fix or reference?
Fixes #1081

### Previous Behavior
The `-R` option wouldn't set up any repo, and would install the default old packages in apt, which is version 0.17. :(

### New Behavior
-R option works well!

```
root@ubuntu:~# sh test.sh -R repo.saltstack.com/staging
<snipped>
root@ubuntu:~# salt-minion --version
salt-minion 2016.11.5 (Carbon)
root@ubuntu:~# cat /etc/apt/sources.list.d/saltstack.list
deb https://repo.saltstack.com/staging/apt/ubuntu/14.04/amd64/latest trusty main
```

I also did a bunch of clean up in this PR that I have been wanting to do for a while. Namely: 
- I moved all of the duplicated warnings about the checking for the correct DPKG_ARCHITECTURE values into a single function
- I created an `install_saltstack_ubuntu_repository` function and moved the saltstack repo set up (and also will work with the `-R` setting with the `_REPO_URL` value) there. This is similar to the way the RHEL distros gate and set up the saltstack repo.
- I created an `install_saltstack_debian_repository` function and moved the saltstack repo set up for BOTH debian 7 and debian 8 into one function. This also works with `-R`, just like the ubuntu repo set up function.
- In these two new functions, I also removed the checks for which "stable" version the user is installing. The packages located on saltstack's repo have been around long enough for now and the old repos are not getting updates any longer.

Commit:
Move dpkg architecture warnings into a single function

Commit:
Create a function to setup the saltstack repo for ubuntu

And remove legacy support for pre-repo.saltstack.com packages for
Ubuntu.

This change also fixes the issue where Ubuntu installations using
the -R option were not working.

Commit:
Typo fix: els --> else

Commit:
Create a function to setup the saltstack repo for debian

This function combines the previous efforts to install the correct
saltstack repo for debian 7 and debian 8 into a single codeblock.

This change also fixes the issue where Debian installations using
the -R option were not working.

Commit:
Spell the function name correctly...

vutny commented on bootstrap-salt.sh:
```diff
@@ -1244,18 +1244,40 @@ __gather_system_info() {
+    __REPO_ARCH="$DPKG_ARCHITECTURE"
+
+    if [ "$DPKG_ARCHITECTURE" = "i386" ]; then
+        echoerror "$_REPO_URL likely doesn't have all required 32-bit packages for Ubuntu $DISTRO_MAJOR_VERSION (yet?)."
```
I guess we should use `${DISTRO_NAME}` variable instead of hard coded `Ubuntu` word here, right?

rallytime replied:
Ah, yes. Good catch. I will fix that.

vutny commented on bootstrap-salt.sh:
```diff
@@ -1244,18 +1244,40 @@ __gather_system_info() {
+        elif [ "$ITYPE" != "git" ]; then
+            echoerror "You can try git installation mode, i.e.: sh ${__ScriptName} git v2016.11.5"
+        fi
+    fi
```
As I see, this should work properly during stable installation on Debian/Raspbian `armhf`, but will output unnecessary error messages which may confuse users.
How about to simplify this whole block to a `case` statement?

rallytime replied:
Good idea.

vutny commented on bootstrap-salt.sh:
```diff
@@ -2887,6 +2886,27 @@ install_ubuntu_check_services() {
+        DEBIAN_CODENAME="wheezy"
+    else
+        DEBIAN_CODENAME="jessie"
+    fi
```
This is fine for now. But I guess in the future we would need a better checks for Debian distros codenames, something like we do for Ubuntu above and in the `__ubuntu_codename_translation` function.

~~Winter~~ Debian Stretch is coming...

rallytime replied:
Yeah, I agree. I thought of this, too, but I think a separate PR will be better for this right now. I'll make an issue for it, especially since 7 is phasing out and we'll be moving to 9 very soon as you said. :)

vutny reviewed (changes requested):
This is super awesome! The code looks DRY and much cleaner now. Great work.

But I would like to offer a couple of suggestions, please take a look below.

s0undt3ch reviewed (approved):
As always, @vutny added valuable input!

Commit:
Use case statement in dpkg check function to simplify error logging

Commit:
Lint fix

rallytime commented:
@vutny I have updated this PR with your suggestions. Please take another look.

vutny commented on bootstrap-salt.sh:
```diff
@@ -1258,24 +1258,30 @@ __check_dpkg_architecture() {
+            error_msg=""
+            ;;
+        "armhf")
+            error_msg=""
```
Since there are no `armhf` packages for Ubuntu and Debian 7, maybe we should also check that here?

vutny commented:
@rallytime That looks really great. Just added some another suggestion above. I know I'm bothersome :smiley:

Commit:
Warn that armhf packages are not available for Ubuntu or Debian 7

- Added checks and warnings for armhf packages not supported on Ubuntu/Debian 7
- Moved __check_dpkg_architecture below __ubuntu_derivatives_translation function
  to use the "DISTRO_NAME_L" value
- Added return codes to various warning cases since if we detect an unsupported
  architecture, we should fail sooner.

rallytime commented:
@vutny Don't worry, this is good feedback! 

My latest change adds warnings for when the architecture detected is `armhf` when installing on Ubuntu and Debian 7. I also moved the `__check_dpkg_architecture` function below the `__ubuntu_derivatives_translation` function to use the "DISTRO_NAME_L" value and added return codes to various warning cases since if we detect an unsupported architecture, we should fail sooner.

Let me know what you think.

Commit:
Just use exit numbers instead of truth values for return codes

rallytime commented:
Go Go Jenkins!

vutny reviewed (approved):
Looks solid!

rallytime commented:
Go Go Jenkins!

rallytime commented:
@s0undt3ch I am not sure why the clone job is failing on these tests. I am not as familiar with this jenkins set up. Do you know why that `ci/clone` build is failing?

Commit:
Lint Fixes
